finality-aleph/src/data_io.rs

```diff
@@ -151,12 +158,16 @@ where
         let requirements: Vec<_> = message
             .included_blocks()
             .into_iter()
-            .filter(|block_hash| {
-                if self.available_blocks.contains(block_hash) {
+            .filter(|block_data| {
+                if self.available_blocks.contains(block_data) {
+                    return false;
+                }
+                if let Ok(Some(_)) = self.client.header(BlockId::Hash(block_data.0)) {
+                    self.add_block(*block_data);
                     return false;
                 }
-                if let Ok(Some(_)) = self.client.header(BlockId::Hash(*block_hash)) {
-                    self.add_block(*block_hash);
+                if self.client.info().finalized_number > block_data.1 {
```

Comment:
It should be even `self.client.info().finalized_number >= block_data.1` -- also in the second place, somewhere above.